zcash_client_backend/src/proposal.rs

```diff
@@ -98,21 +98,127 @@ impl<NoteRef> ShieldedInputs<NoteRef> {
+
+        Ok(Self {
+            fee_rule,
+            min_target_height,
+            steps,
+        })
+    }
+
+    /// Constructs a validated [`Proposal`] having only a single step from its constituent parts.
+    ///
+    /// This operation validates the proposal for balance consistency and agreement between
+    /// the `is_shielding` flag and the structure of the proposal.
+    ///
+    /// Parameters:
+    /// * `transaction_request`: The ZIP 321 transaction request describing the payments to be
+    ///    made.
+    /// * `payment_pools`: A map from payment index to pool type.
+    /// * `transparent_inputs`: The set of previous transparent outputs to be spent.
+    /// * `shielded_inputs`: The sets of previous shielded outputs to be spent.
+    /// * `balance`: The change outputs to be added the transaction and the fee to be paid.
+    /// * `fee_rule`: The fee rule observed by the proposed transaction.
+    /// * `min_target_height`: The minimum block height at which the transaction may be created.
+    /// * `is_shielding`: A flag that identifies whether this is a wallet-internal shielding
+    ///    transaction.
+    #[allow(clippy::too_many_arguments)]
+    pub fn single_step(
+        transaction_request: TransactionRequest,
+        payment_pools: BTreeMap<usize, PoolType>,
+        transparent_inputs: Vec<WalletTransparentOutput>,
+        shielded_inputs: Option<ShieldedInputs<NoteRef>>,
+        balance: TransactionBalance,
+        fee_rule: FeeRuleT,
+        min_target_height: BlockHeight,
+        is_shielding: bool,
+    ) -> Result<Self, ProposalError> {
+        Ok(Self {
+            fee_rule,
+            min_target_height,
+            steps: vec![Step::from_parts(
+                transaction_request,
+                payment_pools,
+                transparent_inputs,
+                shielded_inputs,
+                balance,
+                is_shielding,
+            )?],
+        })
+    }
+
+    /// Returns the fee rule to be used by the transaction builder.
+    pub fn fee_rule(&self) -> &FeeRuleT {
+        &self.fee_rule
+    }
+
+    /// Returns the target height for which the proposal was prepared.
+    ///
+    /// The chain must contain at least this many blocks in order for the proposal to
+    /// be executed.
+    pub fn min_target_height(&self) -> BlockHeight {
+        self.min_target_height
+    }
+
+    /// Returns the steps of the proposal. Each step corresponds to an independent transaction to
+    /// be generated as a result of this proposal.
+    pub fn steps(&self) -> &[Step<NoteRef>] {
+        self.steps.as_ref()
+    }
+}
+
+impl<FeeRuleT: Debug, NoteRef> Debug for Proposal<FeeRuleT, NoteRef> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Proposal")
+            .field("fee_rule", &self.fee_rule)
+            .field("min_target_height", &self.min_target_height)
+            .field("steps", &self.steps)
+            .finish_non_exhaustive()
```

Comment:
This is not non-exhaustive (for `Proposal`) as all of its fields are included in the `Debug` output:
```suggestion
            .finish()
```
Also, I note that `fee_rule` was not included in the previous `Debug` output for `Proposal`; why was that? (Or alternatively, why is it now being included, which is what makes this impl now not non-exhaustive?)

Reply:
I don't know why it wasn't included; it's simple enough to add the `Debug` constraint and I don't see any downside to requiring that. It was probably just an oversight. And for a debug impl, including the fee rule is useful information.